### PR TITLE
Create coarse mesh out of a CGAL::Surface_mesh

### PR DESCRIPTION
**Pull request by fdrmrc**
Given a closed CGAL::Surface_mesh, this function fills its interior with (few) tets and store the resulting 3D tria in the second argument of the present utility function. Starting from that coarse 3D object, one can create a `Quadrature` to integrate over it by exploiting `QGaussSimplex`. 

Therefore in the follow-up PR I'd add the utlity `Quadrature<...>
 compute_global_quadrature_rule(const CGALTtriangulation& tr);`

These together allow to compute quadrature rules over the result of (meaningful) boolean operations between deal.II 3D cells.

### CODE REVIEW

File: include/deal.II/cgal/utilities.h [hunk 1]
```diff
@@ -106,6 +140,36 @@ namespace CGALWrappers
...
+      "The point type of the Surface mesh and the one the 3D mesh are different.");
+    Assert(CGAL::is_closed(sm), ExcMessage("The surface mesh must be closed."));
+
+    using K           = typename CGAL::Kernel_traits<CGALPointType>::Kernel;
+    using Mesh_domain = CGAL::Polyhedral_mesh_domain_with_features_3<
+      K,
+      CGAL::Surface_mesh<CGALPointType>>;
+    using Tr = typename CGAL::Mesh_triangulation_3<Mesh_domain>::type;
+    using Mesh_criteria = CGAL::Mesh_criteria_3<Tr>;
+    using namespace CGAL::parameters;
```
Comment (peterrum) [inline include/deal.II/cgal/utilities.h:163]: Please don't use  `using namespace`.

File: include/deal.II/cgal/utilities.h [hunk 2]
```diff
@@ -64,6 +82,22 @@ namespace CGALWrappers
   template <int dim, typename CGALPointType>
   inline dealii::Point<dim>
   cgal_point_to_dealii_point(const CGALPointType &p);
+
+  /**
+   * Create a coarse mesh out of a closed CGAL::Surface_mesh. This fills the
+   * interior of the surface mesh with tets. This function should be used to get
+   * the coordinates of the (few) tets inside, in order to construct Quadrature
+   * rules over each tetrahedron.
```
[the remark this answers is not in the document]
Comment (fdrmrc) [reply include/deal.II/cgal/utilities.h:90]: I should've been more precise, I'm sorry.
The surface_mesh is indeed what you wrote. With "coarse mesh" I mean a coarse tetrahedralization of the volumetric region bounded by that closed surface.
Once you have this, for instance, you can get vertices of each tetrahedron and compute Quadratures

File: include/deal.II/cgal/utilities.h [hunk 3]
```diff
@@ -64,6 +82,22 @@ namespace CGALWrappers
...
+   *
+   * @param [in] sm The (closed) surface mesh to be filled.
```
[the remark this answers is not in the document]
Comment (fdrmrc) [reply include/deal.II/cgal/utilities.h:92]: You're right. I'll remove the part you highlighted.

File: cmake/config/template-arguments.in [hunk 4]
```diff
@@ -279,6 +279,5 @@ OUTPUT_FLAG_TYPES := { DXFlags; UcdFlags; GnuplotFlags; PovrayFlags; EpsFlags;
                        Deal_II_IntermediateFlags }
 
 // CGAL Kernels
-CGAL_KERNELS := {CGAL::Simple_cartesian<double>; CGAL::Exact_predicates_exact_constructions_kernel;
-                CGAL::Exact_predicates_exact_constructions_kernel_with_sqrt;
```
[the remark this answers is not in the document]
Comment (fdrmrc) [reply cmake/config/template-arguments.in:283]: We don't need that, so I removed it

File: tests/cgal/cgal_surface_mesh_02.cc [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+using C3t3          = CGAL::Mesh_complex_3_in_triangulation_3<Tr,
+                                                     Mesh_domain::Corner_index,
+                                                     Mesh_domain::Curve_index>;
+using namespace CGALWrappers;
+
+void
+test()
+{
+  const std::vector<std::string> fnames{"input_grids/cube.off",
+                                        "input_grids/tetrahedron.off"};
```
Comment (luca-heltai) [inline tests/cgal/cgal_surface_mesh_02.cc:53]: you should also add these two files to the input_grids directory.

File: include/deal.II/cgal/utilities.h [hunk 6]
```diff
@@ -20,10 +20,28 @@
...
+#  include <CGAL/make_mesh_3.h>
+#  include <CGAL/make_surface_mesh.h>
+
+#  include <type_traits>
+
+#  ifdef CGAL_CONCURRENT_MESH_3
+using Concurrency_tag = CGAL::Parallel_tag Concurrency_tag;
+#  else
+using Concurrency_tag = CGAL::Sequential_tag;
+#  endif
```
Comment (luca-heltai) [inline include/deal.II/cgal/utilities.h:44]: Move this inside the CGALWrappers namespace. Same problemw with the space that you had in the test file.

Comment (luca-heltai) [reply include/deal.II/cgal/utilities.h:44]: Also, I'd rename it following deal.II conventions: 

```
using ConcurrencyTag = CGAL::Parallel_tag;
...
```

File: tests/cgal/cgal_surface_mesh_02.cc [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+#include <deal.II/cgal/utilities.h>
+
+#include "../tests.h"
+
+// Create a Surface_mesh from an .off file, then fill it with tets and print
+// vertices
+
+typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
+using CGALPoint = CGAL::Point_3<K>;
+#ifdef CGAL_CONCURRENT_MESH_3
```
Comment (luca-heltai) [inline tests/cgal/cgal_surface_mesh_02.cc:34]: This should become CGALWrappers::ConcurrencyTag

**Comment by fdrmrc** [comment]
Thanks Peter and Luca! I've addressed your comments.

File: include/deal.II/cgal/utilities.h [hunk 8]
```diff
@@ -41,6 +54,12 @@ DEAL_II_NAMESPACE_OPEN
  */
 namespace CGALWrappers
 {
+#  ifdef CGAL_CONCURRENT_MESH_3
+  using Concurrency_tag = CGAL::Parallel_tag;
+#  else
+  using Concurrency_tag = CGAL::Sequential_tag;
```
Comment (luca-heltai) [inline include/deal.II/cgal/utilities.h:60]:
```suggestion
  using ConcurrencyTag = CGAL::Sequential_tag;
```

File: include/deal.II/cgal/utilities.h [hunk 9]
```diff
@@ -106,6 +140,37 @@ namespace CGALWrappers
...
+    using CGALPointType = typename C3t3::Point::Point;
+    Assert(CGAL::is_closed(surface_mesh),
+           ExcMessage("The surface mesh must be closed."));
+
+    using K           = typename CGAL::Kernel_traits<CGALPointType>::Kernel;
+    using Mesh_domain = CGAL::Polyhedral_mesh_domain_with_features_3<
+      K,
+      CGAL::Surface_mesh<CGALPointType>>;
+    using Tr = typename CGAL::
+      Mesh_triangulation_3<Mesh_domain, CGAL::Default, Concurrency_tag>::type;
```
Comment (luca-heltai) [inline include/deal.II/cgal/utilities.h:161]:
```suggestion
      Mesh_triangulation_3<Mesh_domain, CGAL::Default, ConcurrencyTag>::type;
```

File: tests/cgal/cgal_surface_mesh_02.cc [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+#include <CGAL/IO/File_medit.h>
+#include <CGAL/IO/io.h>
+#include <deal.II/cgal/utilities.h>
+
+#include "../tests.h"
+
+// Create a Surface_mesh from an .off file, then fill it with tets and print
+// vertices.
+
+typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
```
Comment (luca-heltai) [inline tests/cgal/cgal_surface_mesh_02.cc:33]:
```suggestion
using K = CGAL::Exact_predicates_inexact_constructions_kernel;
```

File: tests/cgal/cgal_surface_mesh_02.cc [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  C3t3                           tria;
+  unsigned int                   counter = 0;
+  for (const auto &name : fnames)
+    {
+      std::ifstream input(name);
+      input >> sm;
+      cgal_surface_mesh_to_cgal_coarse_triangulation(sm, tria);
+      std::ofstream off_file_medit("coarse.mesh");
+      ++counter;
+      tria.output_to_medit(off_file_medit, false);
```
Comment (luca-heltai) [inline tests/cgal/cgal_surface_mesh_02.cc:61]:
```suggestion
      {
        std::ofstream off_file_medit("coarse.mesh");
        ++counter;
        tria.output_to_medit(off_file_medit, false);
      }
      cat_file("coarse.mesh");
```

File: tests/cgal/cgal_surface_mesh_02.cc [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+      // Loop over tets and print Points
+      deallog << "Vertices of the " + name + " mesh:" << std::endl;
+      for (auto it = tria.cells_in_complex_begin();
+           it != tria.cells_in_complex_end();
+           ++it)
+        {
+          for (unsigned int i = 0; i < 4; ++i)
+            deallog << it->vertex(i)->point() << std::endl;
+        }
```
Comment (luca-heltai) [inline tests/cgal/cgal_surface_mesh_02.cc:70]:
```suggestion
```

File: tests/cgal/cgal_surface_mesh_02.cc [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+      sm.clear(); // reset surface
+      tria.clear();
+      deallog << '\n' << std::endl;
```
Comment (luca-heltai) [inline tests/cgal/cgal_surface_mesh_02.cc:73]:
```suggestion
      deallog << std::endl << std::endl;
```

File: include/deal.II/cgal/utilities.h [hunk 14]
```diff
@@ -64,6 +83,21 @@ namespace CGALWrappers
   template <int dim, typename CGALPointType>
   inline dealii::Point<dim>
   cgal_point_to_dealii_point(const CGALPointType &p);
+
+  /**
+   * Given a closed CGAL::Surface_mesh, this function fills the
+   * internal region bounded by the surface with tets. This should be
+   * used to get the coordinates of the (few) tets inside, in order to construct
+   * Quadrature rules over each tetrahedron.
+   *
```
Comment (luca-heltai) [inline include/deal.II/cgal/utilities.h:92]:
```suggestion
   * Given a closed CGAL::Surface_mesh, this function fills the
   * region bounded by the surface with tets, keeping them as coarse as possible.
   * 
   * The number of the generated tetrahedrons depends on the refinement level of surface mesh. This function
   * does not attempt to construct a fine triangulation, nor to smooth the final result. If you need finer control on 
   * the resulting triangulation, you should consider using directly CGAL::make_mesh_3().
   *
```

File: include/deal.II/cgal/utilities.h [hunk 15]
```diff
@@ -41,6 +54,12 @@ DEAL_II_NAMESPACE_OPEN
  */
 namespace CGALWrappers
 {
+#  ifdef CGAL_CONCURRENT_MESH_3
+  using Concurrency_tag = CGAL::Parallel_tag;
```
[the remark this answers is not in the document]
Comment (luca-heltai) [reply include/deal.II/cgal/utilities.h:58]: Let's keep our notation in our wrappers.

**Review by luca-heltai — CHANGES_REQUESTED** [review]
Just a few more changes for me, and then this is ready from my side. I'll wait for @peterrum to approve it though, since I followed your implementation along the way.

**Comment by peterrum** [comment]
> Just a few more changes for me, and then this is ready from my side. I'll wait for @peterrum to approve it though, since I followed your implementation along the way.

I'll take a look at the most tomorrow. Not sure I get around to it today...

**Comment by fdrmrc** [comment]
> But let me ask you a question: the new function just converts one CGAL class to another one. Is it supposed to be used in combination with one of the other new functions or how should it be used together with deal.II?

It's supposed to be an utility used in combination with the new functions. In practice, once you have this coarse tetrahedralization, you can call
```
template<typename CGALTria>
Quadrature<Tr::Point::Ambient_dimension::value>  compute_quadrature(const CGALTria &tria, const unsigned int degree)
```
which gives a Quadrature rule over the polyhedra. That's still something CGAL-related, to be called inside the following:

```
  template <int dim0, int dim1, int spacedim,>
  dealii::Quadrature<spacedim>
  compute_quadrature_on_boolean_operation(
    const typename dealii::Triangulation<dim0, spacedim>::cell_iterator &cell0,
    const typename dealii::Triangulation<dim1, spacedim>::cell_iterator &cell1,
    const unsigned int                                                   degree,
    const Mapping<dim0, spacedim> &mapping0 =
      (dealii::ReferenceCells::get_hypercube<dim0>()
         .template get_default_linear_mapping<dim0, spacedim>()),
    const Mapping<dim1, spacedim> &mapping1 =
      (dealii::ReferenceCells::get_hypercube<dim1>()
         .template get_default_linear_mapping<dim1, spacedim>()));
```


(I'm going to open a PR for these last two)

**Comment by luca-heltai** [comment]
Also, this together with the boolean operation and another PR that is currently underway, will permit to do the following:

~~~c++
Triangulation<3,3> tria1;
Triangulation<3,3> tria2;
... // Fill tria1, tria2
Triangulation<3,3> tria3;
compute_booloean_operation(tria1,tria2,tria3, bool_op);
~~~
I.e., compute boolean operations on grids, like subtracting balls from cubes, and obtaining a new triangulation in return.

**Review by peterrum — APPROVED** [review]

**Review by luca-heltai — APPROVED** [review]
I'm fine with the current status. Let's go ahead and merge. Another PR is waiting for this.

**Pull request merged**